src/main/java/com/pingcap/tikv/codec/DecimalUtils.java

```diff
@@ -30,18 +30,26 @@
      * @param cdi cdi is source data.
      * */
     public static double readDecimalFully(CodecDataInput cdi) {
-        if (cdi.size() < 3) {
+        if (cdi.available() < 3) {
             throw new IllegalArgumentException("insufficient bytes to read value");
         }
+
+        MyDecimal dec = new MyDecimal();
+        List<Integer> data = new ArrayList<>();
```

Comment:
boxing type in decoding might be too heavy. take a look at guava's http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/primitives/Ints.html#ensureCapacity-int:A-int-int-
to see if can be convert to array of int instead of list of Integer.